text/java/0000-composite-draft-branch.md

```diff
@@ -0,0 +1,39 @@
+# Create a draft branch for composite/meta buildpacks
+
+## Summary
+
+Currently component buildpack bumps are created as PRs against the main branch, e.g. see [here](https://github.com/paketo-buildpacks/java/pull/873).
+This RFC proposes changing the CI to create these dependency PRs against a new branch, e.g. `draft`.    
+
+## Motivation
+
+This will allow us to achieve the following improvements:
+
+1. Auto-merging of component buildpack PRs into this draft branch 
+2. Manual review can be performed with a single PR before merging to main
```

Comment:
It sounds like the main goal of this proposal is to batch the various buildpack bumps to (presumably) reduce the human burden by reviewing fewer, larger changes instead of more, smaller changes.

If so, I don't think you need to open a PR against a `draft` branch. Instead I think you can just push each change to this draft branch (e.g. `buildpack-bumps`) and have a single PR from this branch to `main`. As long as you rebase each change on top of this branch before pushing it, the single PR from `buildpack-bumps` to `main` will contain all the changes, and can be reviewed as a single PR on whatever cadence is useful for the java maintainers. You can run your integration tests on this PR the same as if you opened it manually.

This is how the other Paketo buildpacks work (i.e. the buildpacks that don't use `pipeline-builder`). See [this Python Buildpack PR](https://github.com/paketo-buildpacks/python/pull/555/files) as an example - it contains automated bumps to multiple component buildpacks in a single PR, with automated integration tests. A human can review this PR and manually merge on whatever schedule works for the maintainers of that repository.

I think your proposal will work, and it might reduce the frequency on which Java maintainers review buildpack bumps, but I think you can save yourself even more work if you adopt the flow identified above.

Reply:
Ah, great! Glad we're on the same page.

Reply:
I made a small update to clarify how I intended this to work, as you said @robdimsdale, it would be multiple commits to one draft PR, as per non-Java buildpacks. Thanks!

Reply:
Thanks. I'm glad we're in alignment regarding the high-level goal. The RFC still references a two-step approach of automatic merge of PRs to the `draft` branch followed by a manual merge of PRs to `main`. Maybe I'm missing something, but I think that can be simplified to automatically merging certain PRs directly to `main`. Or at least batching up the changes into a single PR to `main` which is manually merged. This is how it's done in the example you linked to in the Python buildpack.

Do we need the intermediate `draft` branch? Am I missing something?

Reply:
We can't merge directly to main, but this is my understanding:

>Or at least batching up the changes into a single PR to main which is manually merged.

I think `draft` here is just an implementation detail. You need to have a branch to make a PR, so the proposal is calling that `draft`.

Reply:
Right, ok. I think I'm getting hung up on the semantics. In order to "merge" a PR, there already must be a branch open. You're proposing calling that branch `draft` - in the python example lined that branch is called `automation/buildpack.toml/update`. I thought you were proposing opening one branch e.g. `automation`, merging that to `draft`, and then merging that to `main`.

As long as there's only one branch (and hence PR to `main`), I think this is structurally identical to what we have on the other buildpacks.

Reply:
That's correct Rob 👍 sorry for the confusion. Are you happy for me to resolve this thread?